### PR TITLE
Improve UI

### PR DESCRIPTION
**Pull request by AsmardPing**
- Added invite button (that copies URL)
- Added graphical representation of approval rate in result table
- Only shows result table when different options were selected, otherwise consensus cat appears in larger version
- Added Task Control Row and centralised all control functions for the task there
- Moved participants to the bottom of the screen
- Added new Modal for "New Task"
- Re-arranged voting screen to better fit screen dimensions
- Added (limited) mobile screen support
- switched prettier airbnb config from external dependency to direct configuration in prettierrc
- added (limited) connection problem handling
- added key based control to newtaskdialog and showresultdialog

### CODE REVIEW

File: packages/frontend/src/components/EstimationResult.vue [hunk 1]
```diff
@@ -42,13 +61,24 @@
...
     </table>
+  </template>
+
+  <template v-if="consensusReached">
+    <div class="text-xl font-sans m-2 text-center">
+      <span class="font-bold">Consensus Reached! 🎉</span><br /><br />
+      <span class="">{{
+        sortedEntries[0]?.value
+          ? `All ${sortedEntries[0]?.names.length} participants voted for option ${sortedEntries[0]?.value}.`
+          : `All ${sortedEntries[0]?.names.length} participants were to confused to select anything. 🧐`
```
Comment (felixabele) [inline packages/frontend/src/components/EstimationResult.vue]:
```suggestion
          : `All ${sortedEntries[0]?.names.length} participants were too confused to select anything. 🧐`
```

File: packages/frontend/src/components/EstimationResult.vue [hunk 2]
```diff
@@ -78,6 +110,7 @@ const sortFunctions = {
   size: (e1: Entry, e2: Entry): number =>
     cardDeck.value.indexOf(e1.value) - cardDeck.value.indexOf(e2.value),
   votes: (e1: Entry, e2: Entry): number => (e1.names.length < e2.names.length ? -1 : 1),
+  agreement: (e1: Entry, e2: Entry): number => (e1.names.length < e2.names.length ? -1 : 1),
```
Comment (felixabele) [inline packages/frontend/src/components/EstimationResult.vue:112]: Maybe reduce duplication and increase readability, but its just a suggestion:
```
const sortByVotes = (e1: Entry, e2: Entry): number => (e1.names.length < e2.names.length ? -1 : 1);

const sortFunctions = {
  size: (e1: Entry, e2: Entry): number =>
    cardDeck.value.indexOf(e1.value) - cardDeck.value.indexOf(e2.value),
  votes: sortByVotes,
  agreement: sortByVotes
};
```

File: packages/frontend/src/components/OngoingEstimation.vue [hunk 3]
```diff
@@ -1,33 +1,24 @@
...
-      </button>
     </div>
 
-    <div v-if="!isSpectator" class="grid grid-cols-4 gap-x-2 gap-y-2 mb-4">
+    <div v-if="!isSpectator" class="grid grid-cols-5 gap-x-2 gap-y-2 mb-4">
       <card
         v-for="(value, index) in currentCardDeck"
         :ref="
           el => {
+            // @ts-ignore TODO: create proper typing for cardRefList
```
Comment (felixabele) [inline packages/frontend/src/components/OngoingEstimation.vue]: should we solve this TODO in this PR?

File: packages/frontend/src/components/RoomHeader.vue [hunk 4]
```diff
@@ -74,6 +92,13 @@ const leaveRoom = () => {
   router.push({ name: 'lobby', query: { room: props.roomName } });
 };
 
+const urlCopyStatus = ref(false);
+const copyUrl = async () => {
+  //TODO: Shouldn't use plaintext domain here, instead import from wherever it's stored
+  navigator.clipboard.writeText(`https://planningpoker.cc${window.location.pathname}`);
```
Comment (felixabele) [inline packages/frontend/src/components/RoomHeader.vue]: Maybe `window.location.pathname` would be better, so we wouldn't have to hardcode the domain. Or do I miss something?

File: packages/frontend/src/components/TaskHeader.vue [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  <div v-if="taskName" class="w-full min-h-24 relative overflow-hidden box-border">
+    <div class="flex justify-center items-center relative">
+      <div class="w:3/5 flex-1 text-xl font-sans m-2">
+        Task:
+        <span ref="task-name-display" class="font-bold">{{ taskName }} - </span>
+        <span class="">{{ isEstimationOngoing ? 'Estimation' : 'Result' }}</span>
+      </div>
+
+      <div id="controlArea" class="flex justify-end">
+        <button
```
Comment (felixabele) [inline packages/frontend/src/components/TaskHeader.vue]: We should really think about extracting some dump components such as these buttons to DRY code. These long repetitive class concatenations are hard to read.

File: packages/frontend/src/store/mutations.ts [hunk 6]
```diff
@@ -52,12 +54,16 @@ export const mutations: MutationTree<State> & Mutations = {
     state.ongoingEstimation = undefined;
     state.estimationResult = undefined;
   },
+  [MutationsType.ENTER_ROOM](state: State) {
+    state.connectionState = 'CONNECTED';
```
Comment (felixabele) [inline packages/frontend/src/store/mutations.ts:58]: maybe use the enum from `ConnectionState.CONNECTED`

File: packages/frontend/src/store/mutations.ts [hunk 7]
```diff
@@ -52,12 +54,16 @@ export const mutations: MutationTree<State> & Mutations = {
...
+  },
   [MutationsType.LEAVE_ROOM](state: State) {
     state.room = undefined;
     state.cardDeck = [];
     state.participants = [];
     state.ongoingEstimation = undefined;
     state.estimationResult = undefined;
+    state.connectionState = 'NOT_CONNECTED';
```
Comment (felixabele) [inline packages/frontend/src/store/mutations.ts:66]: maybe use the enum from `ConnectionState.NOT_CONNECTED`

File: packages/frontend/src/store/websocketPlugin.ts [hunk 8]
```diff
@@ -64,6 +68,20 @@ const webSocketPlugin = (store: Store<State>) => {
       handleIncomingMessage(json);
     };
 
+    socket.onerror = event => {
+      console.log('[error] Connection error, see log for details.');
+      console.log(event);
```
Comment (felixabele) [inline packages/frontend/src/store/websocketPlugin.ts]:
```suggestion
      console.error('[error] Connection error, see log for details.');
      console.error(event);
```

File: packages/frontend/src/store/websocketPlugin.ts [hunk 9]
```diff
@@ -64,6 +68,20 @@ const webSocketPlugin = (store: Store<State>) => {
...
+    };
+
+    socket.onclose = event => {
+      store.commit('leaveRoom');
+      if (event.wasClean) {
+        console.log(`[close] Connection closed cleanly, code=${event.code}.`);
```
Comment (felixabele) [inline packages/frontend/src/store/websocketPlugin.ts]:
```suggestion
        console.info(`[close] Connection closed cleanly, code=${event.code}.`);
```

File: packages/frontend/src/store/websocketPlugin.ts [hunk 10]
```diff
@@ -64,6 +68,20 @@ const webSocketPlugin = (store: Store<State>) => {
...
+      } else {
+        console.log(`[close] Connection died, code=${event?.code}.`);
```
Comment (felixabele) [inline packages/frontend/src/store/websocketPlugin.ts]:
```suggestion
        console.error(`[close] Connection died, code=${event?.code}.`);
```

File: packages/frontend/src/views/Room.vue [hunk 11]
```diff
@@ -47,11 +55,13 @@ const cardDeck = toRef(store.state, 'cardDeck');
 
 const participants = toRef(store.state, 'participants');
 const taskName = computed(() => {
-  if (typeof store.state.ongoingEstimation === 'undefined') {
+  if (store.state.ongoingEstimation) {
+    return store.state.ongoingEstimation.taskName;
+  } else if (store.state.estimationResult) {
+    return store.state.estimationResult.taskName;
+  } else {
     return '';
```
Comment (felixabele) [inline packages/frontend/src/views/Room.vue]: Maybe less spaghetti but see your self what you like better:
```
const taskName = computed(() => {
  const { ongoingEstimation, estimationResult } = store.state;
  return ongoingEstimation?.taskName || estimationResult?.taskName || ''
});
```

**Review by felixabele — CHANGES_REQUESTED** [review]
Nice work! 

I left some comments, no bugs in the code AFAIC.
I've noticed we have some really complex components and could refactor some by extracting a few dump components to decrease repetitive  markup e.g. buttons or dialog templates. But nothing that blocks this PR.

There is just one weird behaviour I've noticed. You can reproduce with the following steps:
- you have 2 users
- user 1 votes any number
- user 2 votes a different number
- display results
- user 2 leaves
- => both rows stay (which makes sense) but the values both change to 100% now

Vielleicht noch bei EstimationResult einen Test für die Prozent-Anzeige hinzufügen.

File: packages/frontend/src/store/mutations.ts (hunk 6 above)
Comment (AsmardPing) [reply packages/frontend/src/store/mutations.ts:58]: This causes the error "index.ts:38 Uncaught ReferenceError: Cannot access 'mutations' before initialization
    at index.ts:38:3"
    Unfortunately I haven't found a way around it.

File: packages/frontend/src/components/OngoingEstimation.vue (hunk 3 above)
Comment (AsmardPing) [reply packages/frontend/src/components/OngoingEstimation.vue]: I was unsure how to do it. The typescript error was already there beforehand and I didn't implement this part. The error is: 
Type 'Element | ComponentPublicInstance<{}, {}, {}, {}, {}, {}, {}, {}, false, ComponentOptionsBase<any, any, any, any, any, any, any, any, any, {}, {}, string>, {}> | null' is not assignable to type 'VueElement'.
  Type 'null' is not assignable to type 'VueElement'.

File: packages/frontend/src/components/EstimationResult.vue (hunk 2 above)
Comment (AsmardPing) [reply packages/frontend/src/components/EstimationResult.vue:112]: I would personally leave it as it is, as I can see the sort logic of all sort options at one glance right now (even if there is one reptition in it). When starting to outsource it to other methods that would be more effort.

**Comment by AsmardPing** [comment]
@felixabele Bin alles durchgegangen. Das beschriebene Problem mit der sich verändernden Prozentanzeige wenn eine Person den Raum verlässt habe ich behoben. Außerdem alles was "resolved" ist. Ansonsten siehe meine Kommentare :)

File: packages/frontend/src/components/OngoingEstimation.vue (hunk 3 above)
Comment (felixabele) [reply packages/frontend/src/components/OngoingEstimation.vue]: Try with Typecasting from Element to VueElement:
```
          el => {
            cardRefList[index] = el as VueElement;
          }
```

Comment (AsmardPing) [reply packages/frontend/src/components/OngoingEstimation.vue]: it worked :D Could have realised that myself based on the error message actually. Thanks!

**Review by felixabele — APPROVED** [review]

**Comment by felixabele** [comment]
Closed in favour of: https://github.com/Planning-Poker-Teams/planning-poker/pull/48

**Pull request closed without merging**